File: R/bots/posting_policy.R
# ------------------------------------------------------------------------------
# Posting Policy & Cache Helpers for Fourth-Down Bot
# ------------------------------------------------------------------------------

suppressPackageStartupMessages({
  library(dplyr)
  library(data.table)
  library(lubridate)
  library(readr)
})

# Utility: null-coalescing
`%||%` <- function(a, b) if (is.null(a) || is.na(a)) b else a

# ------------------------------------------------------------------------------
# In-memory global state for cooldowns & rate-limiting (reset each R session)
# ------------------------------------------------------------------------------
.post_state <- new.env(parent = emptyenv())
.post_state$last_post_ts_by_game <- list()    # game_id -> POSIXct of last post
.post_state$last_post_drive_by_game <- list() # game_id -> last posted drive_id
.post_state$posts_last_hour_ts <- c()         # POSIXct vector of recent posts

# ------------------------------------------------------------------------------
# Prime/standalone window?
# ------------------------------------------------------------------------------
is_prime_window <- function(game_meta) {
  isTRUE(game_meta$standalone) ||
    (is.character(game_meta$window) &&
       game_meta$window %in% c("MNF","TNF","SNF","HOL","INTL","SAT"))
}

# ------------------------------------------------------------------------------
# Core policy evaluator
# ------------------------------------------------------------------------------
should_post_decision <- function(row, game_meta, now = Sys.time()) {
  # REQUIRED FIELDS in `row`:
  # game_id, drive_id, play_id
  # qtr, sec_left, clock
  # off, def, off_score, def_score, margin
  # ydstogo, yardline_100, down
  # pre_wp, wp_go, wp_punt, wp_fg (0–1 each)
  # best_action, called_action ("go","punt","fg")
  prime <- is_prime_window(game_meta)

  wp_vec <- c(go = row$wp_go, punt = row$wp_punt, fg = row$wp_fg)

  wp_best   <- suppressWarnings(max(wp_vec, na.rm = TRUE))
  wp_called <- wp_vec[[tolower(row$called_action)]]
  wp_gap    <- ifelse(is.na(wp_called), NA_real_, wp_best - wp_called)

  leverage <- wp_best - mean(wp_vec[!is.na(wp_vec)], na.rm = TRUE)

  one_score_2h <- (row$margin <= 8) && (row$qtr >= 3)
  clear_mistake <- (!is.na(wp_gap)) &&
                   (row$best_action != row$called_action) &&
                   (wp_gap >= 0.05)
  high_lev <- !is.na(row$pre_wp) &&
              row$pre_wp >= 0.35 && row$pre_wp <= 0.65 &&
              leverage >= 0.04

  # ---- Global "obvious situation" suppressor ----
  non_na <- wp_vec[!is.na(wp_vec)]
  spread <- if (length(non_na) >= 2) (max(non_na) - min(non_na)) else 0
  obvious_spread_block <- spread >= 0.30              # any options differ by ≥30 pts
  must_go_block <- !is.na(row$wp_punt) && (row$wp_punt < 0.10)  # "must-go" if punt <10%
  obvious_block <- obvious_spread_block || must_go_block

  # ---- Rate limits / blocks ----
  cd_secs <- if (prime) 45 else 60
  last_ts <- .post_state$last_post_ts_by_game[[row$game_id]]
  within_cd <- !is.null(last_ts) &&
               as.numeric(difftime(now, last_ts, units = "secs")) < cd_secs
  cooldown_block <- within_cd && !clear_mistake

  last_drive <- .post_state$last_post_drive_by_game[[row$game_id]]
  drive_block <- !is.null(last_drive) && identical(last_drive, row$drive_id)

  recent <- .post_state$posts_last_hour_ts[
    as.numeric(difftime(now, .post_state$posts_last_hour_ts, units = "mins")) <= 60
  ]
  global_block <- length(recent) >= 12

  blowout <- (row$margin >= 17) && !(row$qtr == 4 && row$sec_left <= 120)
  blowout_block <- blowout && !clear_mistake && (is.na(wp_gap) || wp_gap < 0.08)

  # ---- Eligibility gates by window ----
  reason <- NA_character_
  eligible <-
    if (prime) {
      reason <- "prime_all"; TRUE
    } else if (one_score_2h) {
      reason <- "one_score_2H"; TRUE
    } else if (clear_mistake) {
      reason <- "clear_mistake"; TRUE
    } else if (high_lev) {
      reason <- "high_leverage"; TRUE
    } else {
      FALSE
    }

  # Apply global "obvious" suppressor
  eligible <- eligible && !obvious_block

  post_it <- eligible && !cooldown_block && !drive_block && !global_block && !blowout_block

  list(
    post = post_it,
    reason = ifelse(post_it, reason, NA_character_),
    cooldown_skipped = eligible && within_cd && clear_mistake,
    cooldown_block = cooldown_block,
    drive_block = drive_block,
    global_block = global_block,
    blowout_block = blowout_block,
    # diagnostics
    obvious_block = obvious_block,
    obvious_spread_block = obvious_spread_block,
    must_go_block = must_go_block,
    # metrics
    wp_gap = wp_gap,
    leverage = leverage,
    prime = prime
  )
}

# ------------------------------------------------------------------------------
# Message formatter (PUNT omitted if suppressed/NA)
# ------------------------------------------------------------------------------
.pct <- function(x) ifelse(is.na(x), "", paste0(round(100*as.numeric(x)), "%"))
.yard_str <- function(y100) {
  if (is.na(y100)) return("Own ?")
  if (y100 > 50) paste0("Own ", 100 - y100) else paste0("Opp ", y100)
}

format_post <- function(row, revisionist = FALSE) {
  # quarter label and clock
  qlabel <- if (!is.null(row$qtr) && row$qtr >= 5) "OT" else paste0("Q", row$qtr %||% "?")
  if (!is.null(row$clock) && !is.na(row$clock)) {
    qclock <- row$clock
  } else {
    q <- as.integer(row$qtr %||% 1L)
    s <- as.integer(row$sec_left %||% 0L)
    # seconds remaining in current quarter
    q_left <- s - pmax(0L, (4L - pmin(q, 4L)) * 900L)
    q_left <- max(0L, min(900L, q_left))
    qclock <- sprintf("%d:%02d", q_left %/% 60L, q_left %% 60L)
  }

  # score string
  score_str <- if (!is.null(row$off_score) && !is.null(row$def_score) &&
                   !is.na(row$off_score) && !is.na(row$def_score) &&
                   !is.null(row$off) && !is.null(row$def)) {
    paste0(row$off, " ", row$off_score, " vs ", row$def, " ", row$def_score)
  } else {
    "Score N/A"
  }

  # All-scenarios line: include only available actions; omit PUNT when suppressed/NA
  parts <- character(0)
  if (!is.null(row$wp_fg)   && !is.na(row$wp_fg))   parts <- c(parts, paste0("FG ",  .pct(row$wp_fg)))
  if (!is.null(row$wp_go)   && !is.na(row$wp_go))   parts <- c(parts, paste0("GO ",  .pct(row$wp_go)))
  if (!isTRUE(row$punt_suppressed) && !is.na(row$wp_punt %||% NA_real_)) {
    parts <- c(parts, paste0("PUNT ", .pct(row$wp_punt)))
  }
  all_line <- if (length(parts)) paste(parts, collapse = "  ") else ""

  # recommended action & WP
  ba <- tolower(row$best_action %||% "")
  rec_act <- toupper(ba %||% "—")
  rec_wp  <- {
    key <- paste0("wp_", ba)
    .pct(if (!is.null(row[[key]]) && !is.na(row[[key]])) row[[key]] else row$pre_wp)
  }

  header <- paste0(
    "4th & ", row$ydstogo, " at ", .yard_str(row$yardline_100),
    " | ", score_str, " | ", qlabel, " ", qclock,
    if (!is.null(row$off)) paste0(" (", row$off, " ball)") else ""
  )

  lines <- c(
    header,
    paste0(if (revisionist) "Model Says: " else "Model Recommendation: ",
           rec_act, " (WP ", rec_wp, ")")
  )
  if (nzchar(all_line)) lines <- c(lines, paste0("All Scenarios: ", all_line))

  paste0(paste(lines, collapse = "\n"), "\n#nfl #4thDown #analytics")
}

# ------------------------------------------------------------------------------
# CSV cache appender
# ------------------------------------------------------------------------------
append_cache <- function(row, policy, cache_path) {
  dir.create(dirname(cache_path), recursive = TRUE, showWarnings = FALSE)

  out <- data.frame(
    season = row$season,
    week = row$week,
    game_id = row$game_id,
    drive_id = row$drive_id,
    play_id = row$play_id,
    post_time_utc = format(Sys.time(), "%Y-%m-%d %H:%M:%S", tz = "UTC"),
    posted_flag = policy$post,
    posted_reason = policy$reason,
    qtr = row$qtr,
    clock = row$clock,
    sec_left = row$sec_left,
    off = row$off,
    def = row$def,
    score_off = row$score_off,
    score_def = row$score_def,
    margin = row$margin,
    one_score_2h = (row$margin <= 8 && row$qtr >= 3),
    yardline_100 = row$yardline_100,
    down = row$down,
    ydstogo = row$ydstogo,
    roof = row$roof %||% NA,
    surface = row$surface %||% NA,
    wind = row$wind %||% NA,
    temp = row$temp %||% NA,
    precip = row$precip %||% NA,
    pre_wp = row$pre_wp,
    wp_go = row$wp_go,
    wp_punt = row$wp_punt,
    wp_fg = row$wp_fg,
    best_action = row$best_action,
    called_action = row$called_action,
    wp_gap = policy$wp_gap,
    leverage = policy$leverage,
    prime_window = policy$prime,
    cooldown_skipped = policy$cooldown_skipped,
    global_blocked = policy$global_block,
    notes = row$notes %||% NA_character_,
    stringsAsFactors = FALSE
  )

  is_new <- !file.exists(cache_path)
  readr::write_csv(out, cache_path, append = !is_new)
}

# ------------------------------------------------------------------------------
# Mark a successful post (cooldown / rate limits)
# ------------------------------------------------------------------------------
mark_posted <- function(game_id, drive_id) {
  .post_state$last_post_ts_by_game[[game_id]] <- Sys.time()
  .post_state$last_post_drive_by_game[[game_id]] <- drive_id
  .post_state$posts_last_hour_ts <- c(.post_state$posts_last_hour_ts, Sys.time())
}

# ------------------------------------------------------------------------------
# Main handler (optional helper; uses a project-specific eval_fourth_down())
# ------------------------------------------------------------------------------
handle_fourth_down <- function(gs, models, game_meta) {
  res <- eval_fourth_down(gs, models)  # project-specific

  row <- c(gs, res) %>% as.list()
  row$qtr_label <- paste0("Q", row$qtr)
  row$yardline <- sprintf("%s %d",
                          ifelse(row$yardline_100 > 50, row$off, row$def),
                          100 - row$yardline_100)
  row$score_off <- row$off_score
  row$score_def <- row$def_score
  row$margin   <- abs(row$off_score - row$def_score)

  policy <- should_post_decision(row, game_meta)

  cache_file <- file.path(
    "data/cache", row$season,
    sprintf("week_%02d", as.integer(row$week)),
    sprintf("decisions_%s.csv", row$game_id)
  )
  append_cache(row, policy, cache_file)

  if (isTRUE(policy$post)) {
    is_revisionist <- (!is.na(policy$wp_gap) &&
                       row$best_action != row$called_action &&
                       policy$wp_gap >= 0.05)
    msg <- format_post(row, revisionist = is_revisionist)
    # post_to_social(msg)  # implement in your project
    mark_posted(row$game_id, row$drive_id)
  }

  invisible(policy)
}

File: src/live/posting_policy.py
# src/live/posting_policy.py
# Social Posting Policy & Narrative Formatter for 4th Down Bot.
# References: docs/database_guide.md, R/bots/posting_policy.R
# ------------------------------------------------------------------------------

import logging
from typing import Dict, Any, Tuple
from src.live.state_store import is_game_in_cooldown, is_drive_posted, get_global_posts_last_hour

logger = logging.getLogger("4thDownBot.PostingPolicy")


def should_post_decision(row: Dict[str, Any], game_meta: Dict[str, Any], db_path: str = "data/live/bot_state.db") -> Tuple[bool, str]:
    """
    Evaluates whether a 4th down decision matches our gated posting policies.
    
    Returns:
        A tuple of (should_post: bool, reason: str).
    """
    game_id = str(row.get("game_id"))
    drive_id = str(row.get("drive_id") or f"d_{row.get('play_id')}")
    
    # Extract predicted Win Probabilities
    wp_go = row.get("wp_go")
    wp_punt = row.get("wp_punt")
    wp_fg = row.get("wp_fg")
    
    # Clean and filter NAs from WPs
    wps = {}
    if wp_go is not None and wp_go != "NA":
        wps["go"] = float(wp_go)
    if wp_punt is not None and wp_punt != "NA" and not row.get("punt_suppressed"):
        wps["punt"] = float(wp_punt)
    if wp_fg is not None and wp_fg != "NA" and not row.get("fg_suppressed"):
        wps["fg"] = float(wp_fg)
        
    if not wps:
        return False, "no_valid_predictions"
        
    wp_best = max(wps.values())
    best_action = max(wps, key=wps.get)
    
    called_action = str(row.get("called_action") or "").lower()
    wp_called = wps.get(called_action)
    
    wp_gap = 0.0
    if wp_called is not None:
        wp_gap = wp_best - wp_called
        
    # 1. Gate 1: Mistakes (errors >= 3% WP)
    # Always eligible, always bypasses obviousness blocks.
    clear_mistake = (called_action in wps) and (best_action != called_action) and (wp_gap >= 0.03)
    
    # Calculate Leverage
    leverage = wp_best - (sum(wps.values()) / len(wps))
    
    # 2. Obvious Situation blocks (only suppress correct coaching actions)
    spread = max(wps.values()) - min(wps.values()) if len(wps) >= 2 else 0.0
    obvious_spread_block = spread >= 0.30
    must_go_block = ("punt" in wps) and (wps["punt"] < 0.10)
    obvious_block = obvious_spread_block or must_go_block
    
    # Check Context gates
    off_score = int(row.get("off_score") or 0)
    def_score = int(row.get("def_score") or 0)
    margin = abs(off_score - def_score)
    qtr = int(row.get("qtr") or 1)
    
    one_score_game = margin <= 8
    
    # Check Standalone Primetime (Gate 3)
    # Passed via game_meta from main daemon tracking active scoreboards
    is_standalone_prime = bool(game_meta.get("is_standalone_prime", False))
    
    # Eligibility checks
    eligible = False
    reason = "none"
    
    if clear_mistake:
        eligible = True
        reason = "clear_mistake"
    elif is_standalone_prime:
        # Standalone gets coverage for all non-obvious plays
        if not obvious_block:
            eligible = True
            reason = "prime_standalone"
    elif one_score_game:
        # Standard one-possession leverage coverage
        if not obvious_block and leverage >= 0.04:
            eligible = True
            reason = "one_score_leverage"
            
    if not eligible:
        return False, "not_eligible"
        
    # 3. Cooldowns & Rate Limits
    # Mistakes bypass the standard game-level cooldown to post instantly
    is_masto_bsky_prime = is_standalone_prime or game_meta.get("window") in ("MNF", "TNF", "SNF")
    cd_secs = 45 if is_masto_bsky_prime else 60
    
    if is_game_in_cooldown(game_id, cd_secs, db_path=db_path) and not clear_mistake:
        return False, "game_cooldown_block"
        
    # Drive block is absolute (prevents duplicate plays on same drive)
    if is_drive_posted(game_id, drive_id, db_path=db_path):
        return False, "drive_already_posted"
        
    # Global hourly rate limit is absolute (max 12 posts per hour)
    if get_global_posts_last_hour(db_path=db_path) >= 12:
        return False, "global_rate_limit_block"
        
    return True, reason


def format_post(row: Dict[str, Any], revisionist: bool = False) -> str:
    """
    Formats the textual post narrative for Mastodon and Bluesky.
    """
    qtr = int(row.get("qtr") or 1)
    q_label = "OT" if qtr >= 5 else f"Q{qtr}"
    clock = row.get("clock") or "?:??"
    
    off = row.get("off") or "Offense"
    def_team = row.get("def") or "Defense"
    off_score = row.get("off_score")
    def_score = row.get("def_score")
    
    score_str = f"{off} {off_score} vs {def_team} {def_score}" if off_score is not None and def_score is not None else "Score N/A"
    
    # Format Win Probabilities
    parts = []
    wp_fg = row.get("wp_fg")
    wp_go = row.get("wp_go")
    wp_punt = row.get("wp_punt")
    
    if wp_fg is not None and wp_fg != "NA" and not row.get("fg_suppressed"):
        parts.append(f"FG {round(float(wp_fg)*100)}%")
    if wp_go is not None and wp_go != "NA":
        parts.append(f"GO {round(float(wp_go)*100)}%")
    if wp_punt is not None and wp_punt != "NA" and not row.get("punt_suppressed"):
        parts.append(f"PUNT {round(float(wp_punt)*100)}%")
        
    all_line = "  ".join(parts)
    
    yardline_100 = row.get("yardline_100")
    if yardline_100 is None or yardline_100 == "NA":
        yard_str = "Own ?"
    else:
        y100 = int(float(yardline_100))
        yard_str = f"Opp {y100}" if y100 <= 50 else f"Own {100 - y100}"
        
    header = f"4th & {row.get('ydstogo')} at {yard_str} | {score_str} | {q_label} {clock} ({off} ball)"
    
    if revisionist:
        called = str(row.get("called_action") or "go").lower()
        best = str(row.get("best_action") or "go").lower()
        
        # WP values
        wp_called_val = row.get(f"wp_{called}")
        wp_best_val = row.get(f"wp_{best}")
        
        called_pct = f"{round(float(wp_called_val)*100)}%" if wp_called_val is not None and wp_called_val != "NA" else "N/A"
        best_pct = f"{round(float(wp_best_val)*100)}%" if wp_best_val is not None and wp_best_val != "NA" else "N/A"
        
        delta_pct = "N/A"
        if wp_best_val is not None and wp_best_val != "NA" and wp_called_val is not None and wp_called_val != "NA":
            delta_pct = f"{round(abs(float(wp_best_val) - float(wp_called_val))*100)}%"
            
        action_names = {"go": "Go For It", "fg": "Field Goal", "punt": "Punt"}
        called_name = action_names.get(called, called.upper())
        best_name = action_names.get(best, best.upper())
        
        correct_str = "CORRECT" if called == best else "INCORRECT"
        
        summary_line = (
            f"The {off} made the {correct_str} decision. "
            f"{called_name} gave them a {called_pct} win probability, "
            f"whereas {best_name} would have given them {best_pct} "
            f"(a delta of {delta_pct})."
        )
    else:
        best = str(row.get("best_action") or "go").lower()
        wp_best_val = row.get(f"wp_{best}")
        best_pct = f"{round(float(wp_best_val)*100)}%" if wp_best_val is not None and wp_best_val != "NA" else "N/A"
        rec_act = best.upper()
        summary_line = f"Model Recommendation: {rec_act} (WP {best_pct})"
        
    lines = [header, summary_line]
    if all_line:
        lines.append(f"All Scenarios: {all_line}")
        
    return "\n".join(lines) + "\n#nfl #4thDown #analytics"
